File: src/guillotine-go/primitives/bytes.go
package primitives

import (
	"encoding/hex"
	"fmt"
	"strings"
)

// Bytes represents a variable-length byte array
type Bytes struct {
	data []byte
}

// NewBytes creates a new Bytes from a byte slice
func NewBytes(data []byte) Bytes {
	// Create a copy to avoid external modifications
	result := make([]byte, len(data))
	copy(result, data)
	return Bytes{data: result}
}

// BytesFromHex creates Bytes from a hex string
func BytesFromHex(s string) (Bytes, error) {
	// Remove 0x prefix if present
	s = strings.TrimPrefix(s, "0x")
	
	// Add leading zero if odd length
	if len(s)%2 != 0 {
		s = "0" + s
	}
	
	data, err := hex.DecodeString(s)
	if err != nil {
		return Bytes{}, fmt.Errorf("invalid hex bytes: %w", err)
	}
	
	return NewBytes(data), nil
}

// EmptyBytes creates an empty Bytes
func EmptyBytes() Bytes {
	return Bytes{data: []byte{}}
}

// Data returns a copy of the underlying byte data
func (b Bytes) Data() []byte {
	result := make([]byte, len(b.data))
	copy(result, b.data)
	return result
}

// Len returns the length of the byte array
func (b Bytes) Len() int {
	return len(b.data)
}

// IsEmpty returns true if the byte array is empty
func (b Bytes) IsEmpty() bool {
	return len(b.data) == 0
}

// Hex returns the bytes as a hex string with 0x prefix
func (b Bytes) Hex() string {
	if len(b.data) == 0 {
		return "0x"
	}
	return "0x" + hex.EncodeToString(b.data)
}

// String returns the bytes as a hex string with 0x prefix
func (b Bytes) String() string {
	return b.Hex()
}

// Equal returns true if two Bytes are equal
func (b Bytes) Equal(other Bytes) bool {
	if len(b.data) != len(other.data) {
		return false
	}
	
	for i, v := range b.data {
		if v != other.data[i] {
			return false
		}
	}
	
	return true
}

// Append creates a new Bytes with additional data appended
func (b Bytes) Append(data []byte) Bytes {
	result := make([]byte, len(b.data)+len(data))
	copy(result, b.data)
	copy(result[len(b.data):], data)
	return Bytes{data: result}
}

// Slice returns a new Bytes containing a slice of the data
func (b Bytes) Slice(start, end int) (Bytes, error) {
	if start < 0 || end < 0 || start > end || end > len(b.data) {
		return Bytes{}, fmt.Errorf("invalid slice bounds: start=%d, end=%d, len=%d", start, end, len(b.data))
	}
	
	result := make([]byte, end-start)
	copy(result, b.data[start:end])
	return Bytes{data: result}, nil
}

// At returns the byte at the given index
func (b Bytes) At(index int) (byte, error) {
	if index < 0 || index >= len(b.data) {
		return 0, fmt.Errorf("index out of bounds: %d", index)
	}
	return b.data[index], nil
}

// MarshalText implements encoding.TextMarshaler
func (b Bytes) MarshalText() ([]byte, error) {
	return []byte(b.Hex()), nil
}

// UnmarshalText implements encoding.TextUnmarshaler
func (b *Bytes) UnmarshalText(text []byte) error {
	bytes, err := BytesFromHex(string(text))
	if err != nil {
		return err
	}
	*b = bytes
	return nil
}

File: src/guillotine-ts/src/primitives/bytes.ts
/**
 * Represents a variable-length byte array
 */
export class Bytes {
  private readonly _data: Uint8Array;

  private constructor(data: Uint8Array) {
    this._data = new Uint8Array(data);
  }

  /**
   * Creates Bytes from a byte array
   */
  static fromBytes(data: Uint8Array | ArrayLike<number>): Bytes {
    return new Bytes(new Uint8Array(data));
  }

  /**
   * Creates Bytes from a hex string
   */
  static fromHex(hex: string): Bytes {
    // Remove 0x prefix if present
    const cleanHex = hex.startsWith('0x') ? hex.slice(2) : hex;
    
    // Handle empty hex
    if (cleanHex.length === 0) {
      return new Bytes(new Uint8Array(0));
    }
    
    // Pad with leading zero if odd length
    const paddedHex = cleanHex.length % 2 === 0 ? cleanHex : '0' + cleanHex;
    
    if (!/^[0-9a-fA-F]*$/.test(paddedHex)) {
      throw new Error('Invalid hex bytes: contains non-hex characters');
    }
    
    const bytes = new Uint8Array(paddedHex.length / 2);
    for (let i = 0; i < bytes.length; i++) {
      bytes[i] = parseInt(paddedHex.slice(i * 2, i * 2 + 2), 16);
    }
    
    return new Bytes(bytes);
  }

  /**
   * Creates Bytes from a string (UTF-8 encoded)
   */
  static fromString(str: string): Bytes {
    const encoder = new TextEncoder();
    return new Bytes(encoder.encode(str));
  }

  /**
   * Creates empty Bytes
   */
  static empty(): Bytes {
    return new Bytes(new Uint8Array(0));
  }

  /**
   * Creates Bytes with the specified length, filled with zeros
   */
  static zeros(length: number): Bytes {
    if (length < 0) {
      throw new Error('Length cannot be negative');
    }
    return new Bytes(new Uint8Array(length));
  }

  /**
   * Creates Bytes with the specified length, filled with the given value
   */
  static filled(length: number, value: number): Bytes {
    if (length < 0) {
      throw new Error('Length cannot be negative');
    }
    if (value < 0 || value > 255) {
      throw new Error('Value must be between 0 and 255');
    }
    return new Bytes(new Uint8Array(length).fill(value));
  }

  /**
   * Returns a copy of the underlying byte data
   */
  toBytes(): Uint8Array {
    return new Uint8Array(this._data);
  }

  /**
   * Returns the length of the byte array
   */
  length(): number {
    return this._data.length;
  }

  /**
   * Returns true if the byte array is empty
   */
  isEmpty(): boolean {
    return this._data.length === 0;
  }

  /**
   * Returns the bytes as a hex string with 0x prefix
   */
  toHex(): string {
    if (this._data.length === 0) {
      return '0x';
    }
    return '0x' + Array.from(this._data)
      .map(byte => byte.toString(16).padStart(2, '0'))
      .join('');
  }

  /**
   * Returns the bytes as a UTF-8 decoded string
   */
  toString(): string {
    const decoder = new TextDecoder();
    return decoder.decode(this._data);
  }

  /**
   * Returns true if two Bytes are equal
   */
  equals(other: Bytes): boolean {
    if (this._data.length !== other._data.length) {
      return false;
    }
    
    for (let i = 0; i < this._data.length; i++) {
      if (this._data[i] !== other._data[i]) {
        return false;
      }
    }
    
    return true;
  }

  /**
   * Concatenates this Bytes with another Bytes
   */
  concat(other: Bytes): Bytes {
    const combined = new Uint8Array(this._data.length + other._data.length);
    combined.set(this._data, 0);
    combined.set(other._data, this._data.length);
    return new Bytes(combined);
  }

  /**
   * Returns a slice of the bytes
   */
  slice(start: number, end?: number): Bytes {
    if (start < 0 || start > this._data.length) {
      throw new Error(`Start index out of bounds: ${start}`);
    }
    
    const actualEnd = end ?? this._data.length;
    if (actualEnd < start || actualEnd > this._data.length) {
      throw new Error(`End index out of bounds: ${actualEnd}`);
    }
    
    return new Bytes(this._data.slice(start, actualEnd));
  }

  /**
   * Returns a specific byte at the given index
   */
  at(index: number): number {
    if (index < 0 || index >= this._data.length) {
      throw new Error(`Index out of bounds: ${index}`);
    }
    return this._data[index];
  }

  /**
   * Sets a specific byte at the given index (returns new Bytes)
   */
  setAt(index: number, value: number): Bytes {
    if (index < 0 || index >= this._data.length) {
      throw new Error(`Index out of bounds: ${index}`);
    }
    if (value < 0 || value > 255) {
      throw new Error('Value must be between 0 and 255');
    }
    
    const newData = new Uint8Array(this._data);
    newData[index] = value;
    return new Bytes(newData);
  }

  /**
   * Pads the bytes to the specified length with zeros on the left
   */
  padStart(targetLength: number): Bytes {
    if (targetLength <= this._data.length) {
      return new Bytes(this._data);
    }
    
    const padded = new Uint8Array(targetLength);
    padded.set(this._data, targetLength - this._data.length);
    return new Bytes(padded);
  }

  /**
   * Pads the bytes to the specified length with zeros on the right
   */
  padEnd(targetLength: number): Bytes {
    if (targetLength <= this._data.length) {
      return new Bytes(this._data);
    }
    
    const padded = new Uint8Array(targetLength);
    padded.set(this._data, 0);
    return new Bytes(padded);
  }

  /**
   * Finds the index of the first occurrence of the given byte
   */
  indexOf(value: number): number {
    if (value < 0 || value > 255) {
      throw new Error('Value must be between 0 and 255');
    }
    return this._data.indexOf(value);
  }

  /**
   * Returns true if the bytes contain the given value
   */
  includes(value: number): boolean {
    return this.indexOf(value) !== -1;
  }

  /**
   * Reverses the byte order (returns new Bytes)
   */
  reverse(): Bytes {
    const reversed = new Uint8Array(this._data);
    reversed.reverse();
    return new Bytes(reversed);
  }

  /**
   * JSON serialization
   */
  toJSON(): string {
    return this.toHex();
  }

  /**
   * Create from JSON
   */
  static fromJSON(json: string): Bytes {
    return Bytes.fromHex(json);
  }
}